### PR TITLE
Autofill sequence metadata from the submission start forms

### LINKED ISSUE #1208: Submission start - Autofill the sequence metadata
### Inspiration

The new submission start page submits `operator`, `protocol`, `instrument`, and `date` associated with each peak annotation file.

### Description

Autofill that information into the `Sequences` sheet and autofill Sequence Names into the `Peak Annotation Files` and `Peak Annotation Details` sheets.

### Alternatives

None

### Dependencies

This issue cannot be started until the completion of the following issue(s)/
pull request(s):

- #1205

### Comment

None

-----

### Assumptions

1. List of assumptions that the code will not explicitly address/check
2. E.g. We will assume input is correct (explaining why there is no validation)

### Limitations

1. A list of things this work will specifically not do
2. E.g. This feature will only handle the most frequent use case X

### Affected Components

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

#### Interface Change description

None provided

#### Code Change Description

None provided

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

This adds processing to the `BuildSubmissionView`'s form, now that it was operator, instrument, protocol, and run date, to autofill that data into the downloaded template.

## Affected Issues/Pull Requests

- Resolves #1208
- Merges into main

## Review Notes

See comments in-line.

## Checklist

This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    
  - Review period: 2 days 
- Associated issue/pull request requirements:
  
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

### CODE REVIEW

File: DataRepo/loaders/sequences_loader.py [hunk 1]
```diff
@@ -147,7 +147,7 @@ class SequencesLoader(TableLoader):
                 f'NOT(ISBLANK(INDIRECT("{{{INSTRUMENT_KEY}}}" & ROW()))),'
                 f'NOT(ISBLANK(INDIRECT("{{{DATE_KEY}}}" & ROW()))),'
                 "),"
-                f'_xlfn.TEXTJOIN("{MSRunSequence.SEQNAME_DELIMITER} ", FALSE, '
+                f'_xlfn.TEXTJOIN("{MSRunSequence.SEQNAME_DELIMITER}", FALSE, '
```
Comment (hepcat72) [inline DataRepo/loaders/sequences_loader.py:150]: C. I moved the added space to the class delimiter attribute in `MSRunSequence` so that everything would be consistent.

File: DataRepo/models/msrun_sequence.py [hunk 2]
```diff
@@ -132,6 +134,38 @@ def parse_sequence_name(cls, sequence_name: str):
             date,
         )
```
Comment (hepcat72) [inline DataRepo/models/msrun_sequence.py:136]: C. Added the method below for consistency/re-use.

File: DataRepo/views/upload/submission.py [hunk 3]
```diff
@@ -263,6 +263,7 @@ def __init__(self):
         self.annot_files_dict: Dict[str, str] = {}
         self.peak_annot_files = None
         self.peak_annot_filenames = []
+        self.annot_file_metadata = {}
```
Comment (hepcat72) [inline DataRepo/views/upload/submission.py:266]: C. This is where the sequence metadata from the form will be saved.

File: DataRepo/views/upload/submission.py [hunk 4]
```diff
@@ -607,45 +608,64 @@ def get_context_data(self, **kwargs):
...
-
-        self.load_status_data.clear_load()
-
-        self.set_files(
-            tmp_study_file,
-            study_filename=str(study_doc) if study_doc is not None else None,
-            peak_annot_files=[fp.temporary_file_path() for fp in peak_annotation_files],
-            peak_annot_filenames=[str(fp) for fp in peak_annotation_files],
-        )
-
```
Comment (hepcat72) [inline DataRepo/views/upload/submission.py:641]: C. Moving the formset processing to `form_valid` makes access of the data cleaner.

File: DataRepo/views/upload/submission.py [hunk 5]
```diff
@@ -607,45 +608,64 @@ def get_context_data(self, **kwargs):
...
+                peak_annot_files.append(peak_annot_file)
+                peak_annot_filenames.append(peak_annot_filename)
+
+                self.annot_file_metadata[peak_annot_filename] = {
+                    "operator": rowform.get("operator"),
+                    "protocol": rowform.get("protocol"),
+                    "instrument": rowform.get("instrument"),
+                    # Date key is selected to be compatible with MSRunSequence.create_sequence_name
+                    "date": rowform.get("run_date"),
+                }
```
Comment (hepcat72) [inline DataRepo/views/upload/submission.py:658]: C. Cleaner access of formset data instead of messing with prefixes that django automatically removes.

File: DataRepo/views/upload/submission.py [hunk 6]
```diff
@@ -672,7 +692,6 @@ def form_valid(self, formset):
         #       'run_date': '2024-09-18',
         #     },
         #   ]
-        mode = formset.cleaned_data[0]["mode"]
```
Comment (hepcat72) [inline DataRepo/views/upload/submission.py:675]: C. Moved to above.

File: DataRepo/views/upload/submission.py [hunk 7]
```diff
@@ -766,7 +785,7 @@ def get_download_data(self):
             print("Extracting autofill information from file")
 
             # Extract autofill data directly from the peak annotation files
-            self.extract_autofill_from_files()
+            self.extract_autofill_from_files_and_forms()
```
Comment (hepcat72) [inline DataRepo/views/upload/submission.py:788]: C. Renamed a couple methods for accuracy.

File: DataRepo/views/upload/submission.py [hunk 8]
```diff
@@ -1218,10 +1237,11 @@ def add_formulas(self, xlsx_writer):
                             f"{col_letter}{rownum}", formula, None, ""
                         )
 
-    def extract_autofill_from_files(self):
-        """Calls methods for extracting autofill data from the submitted files."""
+    def extract_autofill_from_files_and_forms(self):
+        """Calls methods for extracting autofill data from the submitted forms (including from the submitted files)."""
```
Comment (hepcat72) [inline DataRepo/views/upload/submission.py:1241]: C. This method is not yet used, but will be once the ability to "add to an existing study doc" is restored.  We removed it awhile back, but Michael said it would be useful, and it is a part of the phase 2 plan, as laid out in the discussion in #1040.

File: DataRepo/views/upload/submission.py [hunk 9]
```diff
@@ -1531,14 +1595,53 @@ def extract_autofill_from_peak_annot_details_sheet(self):
             seq_name = loader.get_row_val(row, loader.headers.SEQNAME)
             self.extract_autofill_from_sequence_name(seq_name, seen)
```
Comment (hepcat72) [inline DataRepo/views/upload/submission.py:1597]: C. The code below represents a special case.  Normally, these "extract autofill" methods take data out of the sheet in question and populate it into the autofill dict to be put in other sheets, but that MO relies on there being enough information in the sheet undergoing extraction to uniquely identify a unique row of where it will be used for autofill.  Instead, we have sequence info and the file associated with it, which was extracted from the `Peak Annotation Files` sheet, and we want to fill in the sequence name in this sheet (`Peak Annotation Details`).  This sheet defines a unique row by file and sample header, so in order to populate the sequence in the autofill dict, we need the sample names taken from this sheet.  The data structure looks like this:

```
unique_row_key = delimiter.join([sample_header, filename])
self.autofill_dict[sheet_name][unique_row_key][column_name] = value
```

That's why the autofill is populated here instead of the extraction method for the peak annotation files sheet.

The autofill mechanism could probably be improved to support non-unique row identification, if we're always filling in the same thing on multiple rows...  but that's not the purpose of this PR.

File: DataRepo/widgets.py [hunk 10]
```diff
@@ -70,11 +70,9 @@ def datalist(self):
                 vals += f"<option>{val}</option>"
             elif isinstance(val, tuple):
                 if val[0].lower().replace(" ", "") == val[1].lower().replace(" ", ""):
-                    vals += f"<option>{val[1]}</option>"
+                    vals += f"<option>{val[0]}</option>"
                 else:
-                    vals += (
-                        f"<option data-value='{val[0]}'>{val[0]} ({val[1]})</option>"
-                    )
+                    vals += f"<option value='{val[0]}'>{val[1]}</option>"
```
Comment (hepcat72) [inline DataRepo/widgets.py:75]: C. With this effort, I realized that datalist selections didn't work how I'd interpreted what I'd read about the data-value attribute.  By setting the `value` attribute, THAT is what ends up being submitted.

File: DataRepo/models/msrun_sequence.py [hunk 11]
```diff
@@ -132,6 +134,38 @@ def parse_sequence_name(cls, sequence_name: str):
...
 
+    @classmethod
+    def create_sequence_name(
```
Comment (lparsons) [inline DataRepo/models/msrun_sequence.py:138]: C: Not needed, but for consistency, I'd suggest creating an instance method that uses this method and calling it `sequence_name`. It could be a property as well. Those two methods would probably be sufficient for a separate PR if desired.

File: DataRepo/views/upload/submission.py [hunk 12]
```diff
@@ -1531,14 +1595,53 @@ def extract_autofill_from_peak_annot_details_sheet(self):
...
 
+            # We can actually fill in any missing seqnames in this sheet based on the file (and sample header)
+            # columns...
+            sample_header = loader.get_row_val(row, loader.headers.SAMPLEHEADER)
+            file = loader.get_row_val(row, loader.headers.ANNOTNAME)
+            if file in self.none_vals or sample_header in self.none_vals:
+                continue
+            filename = (os.path.split(file))[1]
```
Comment (lparsons) [inline DataRepo/views/upload/submission.py:1604]: Probably clearer to use [`os.path.basename()`](https://docs.python.org/3/library/os.path.html#os.path.basename) here.
```suggestion
            filename = os.path.basename(file)
```

**Review by lparsons — APPROVED** [review]
Just a couple of minor notes on this PR. Looks good to merge.

File: DataRepo/models/msrun_sequence.py (hunk 11 above)
Comment (hepcat72) [reply DataRepo/models/msrun_sequence.py:138]: Oh wait... no, I need this to be a class method, to create names before an object exists...

Comment (lparsons) [reply DataRepo/models/msrun_sequence.py:138]:
```python
    @classmethod
    def create_sequence_name( ... )
        ...
        return name

    def sequence_name(self) {
        return create_sequence_name(self.operator, self.protocol, self.instrument, self.date)
    }
```

Comment (hepcat72) [reply DataRepo/models/msrun_sequence.py:138]: `git grep -A 2 create_sequence_name DataRepo/views/upload/submission.py`
```
                    ] = MSRunSequence.create_sequence_name(
                        **self.annot_file_metadata[filename]
                    )

...

            seqname = MSRunSequence.create_sequence_name(**seq_data)

...

                partial_seqname = MSRunSequence.create_sequence_name(**seq_data)
```

Comment (hepcat72) [reply DataRepo/models/msrun_sequence.py:138]: Oh, I see what you're saying... I don't think that there's a need for an instance method.  It's only used in the submission template.
